Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift

```diff
@@ -114,66 +148,36 @@ func _call_host_function_impl(
     _ = callbackFuncRef(result)
 }
 
-// MARK: - Legacy Closure Types
 
-/// `JSOneshotClosure` is a JavaScript function that can be called only once.
-/// It is recommended to use `JSClosure` instead if your target runtimes support `FinalizationRegistry`.
-public class JSOneshotClosure: JSObject, JSClosureProtocol {
-    private var hostFuncRef: JavaScriptHostFuncRef = 0
 
-    public init(_ body: @escaping ([JSValue]) -> JSValue) {
-        // 1. Fill `id` as zero at first to access `self` to get `ObjectIdentifier`.
-        super.init(id: 0)
-        let objectId = ObjectIdentifier(self)
-        let funcRef = JavaScriptHostFuncRef(bitPattern: Int32(objectId.hashValue))
-        // 2. Retain the given body in static storage by `funcRef`.
-        sharedClosures[funcRef] = {
-            defer { self.release() }
-            return body($0)
-        }
-        // 3. Create a new JavaScript function which calls the given Swift function.
-        var objectRef: JavaScriptObjectRef = 0
-        _create_function(funcRef, &objectRef)
+// [WeakRefs](https://github.com/tc39/proposal-weakrefs) is already Stage 4, but it's still newish API,
+// so provide a escape hatch.
+// Please build with `-Xswiftc -DJAVASCRIPTKIT_WITHOUT_WEAKREFS` by SwiftPM build system
+#if JAVASCRIPTKIT_WITHOUT_WEAKREFS
 
-        hostFuncRef = funcRef
-        id = objectRef
-    }
+// MARK: - Legacy Closure Types
 
-    /// Release this function resource.
-    /// After calling `release`, calling this function from JavaScript will fail.
+extension JSClosure {
     public func release() {
-        sharedClosures[hostFuncRef] = nil
+        isReleased = true
+        Self.sharedClosures[hostFuncRef] = nil
     }
 }
 
-public class JSUnretainedClosure: JSObject, JSClosureProtocol {
-    private var hostFuncRef: JavaScriptHostFuncRef = 0
-    var isReleased: Bool = false
+@_cdecl("_free_host_function_impl")
+func _free_host_function_impl(_ hostFuncRef: JavaScriptHostFuncRef) {}
 
-    public init(_ body: @escaping ([JSValue]) -> JSValue) {
-        // 1. Fill `id` as zero at first to access `self` to get `ObjectIdentifier`.
-        super.init(id: 0)
-        let objectId = ObjectIdentifier(self)
-        let funcRef = JavaScriptHostFuncRef(bitPattern: Int32(objectId.hashValue))
-        // 2. Retain the given body in static storage by `funcRef`.
-        sharedClosures[funcRef] = body
-        // 3. Create a new JavaScript function which calls the given Swift function.
-        var objectRef: JavaScriptObjectRef = 0
-        _create_function(funcRef, &objectRef)
+#else
 
-        hostFuncRef = funcRef
-        id = objectRef
-    }
+extension JSClosure {
 
-    public func release() {
-        isReleased = true
-        sharedClosures[hostFuncRef] = nil
-    }
+    @available(*, deprecated, message: "JSClosure.release() is no longer necessary if the target environment supports WeakRefs")
```

Comment:
IMO this deprecation message implies some sort of feature checking has happened to determine that WeakRefs are supported.

```suggestion
    @available(*, deprecated, message: "JSClosure.release() is no longer necessary")
```